src/modm/platform/core/rp/multicore.hpp

```diff
@@ -0,0 +1,184 @@
+	 */
+	struct Mailbox {
+		static inline bool canRead()
+		{
+			return !!(sio_hw->fifo_st & SIO_FIFO_ST_VLD_BITS);
+		}
+		static inline bool canWrite()
+		{
+			return !!(sio_hw->fifo_st & SIO_FIFO_ST_RDY_BITS);
+		}
+		static inline bool push(uint32_t msg)
+		{
+			if (!canWrite()) return false;
+			sio_hw->fifo_wr = msg;
+			__SEV();
+			return true;
+		}
+		static inline bool pop(uint32_t& msg)
+		{
+			if (!canRead()) return false;
+			msg = sio_hw->fifo_rd;
+			return true;
+		}
+		static inline void pushBlocked(uint32_t msg)
+		{
+			while (!canWrite()) __NOP();
+			sio_hw->fifo_wr = msg;
+			__SEV();
+		}
+		static inline uint32_t popBlocked()
+		{
+			while (!canRead()) __WFE();
+			return sio_hw->fifo_rd;
+		}
+		static void drain()
+		{
+			while (canRead())
+				(void)sio_hw->fifo_rd;
+		}
+	};
+
+	/**
+	 * The RP2040 provides 32 hardware spin locks, which can be used to manage mutually-exclusive access to shared software
+ 	 * and hardware resources.
+ 	 *
+ 	 * Instance 0 allocated for modm usage.
+ 	 *
+ 	 * Acquire a spin lock without disabling interrupts (hence unsafe)
+ 	 */
+	template <uint8_t instance>
+	class SpinLockUnsafe {
+		static_assert(instance < 32);
+		using LockType = volatile uint32_t;
+	private:
+		static inline LockType* getInstance() {
+			return (LockType *) (SIO_BASE + SIO_SPINLOCK0_OFFSET + instance * 4);
+		}
+	public:
+		static void init() {
+			unlock();
+		}
+		static void lock() {
+			auto l = getInstance();
+			while (*l == 0);
+			std::atomic_thread_fence(std::memory_order_acquire);
+		}
+		static void unlock() {
+			auto l = getInstance();
+			std::atomic_thread_fence(std::memory_order_release);
+			*l = 0;
+		}
+	};
+
+	/**
+	 *
+	 * This class will disable interrupts prior to acquiring the spinlock
+	 */
+	template <uint8_t instance>
+	class SpinLockBlocking : public SpinLockUnsafe<instance>{
+		using Base = SpinLockUnsafe<instance>;
+	public:
+		static uint32_t lock() {
+			uint32_t is = __get_PRIMASK();
+			__disable_irq();
+			Base::lock();
+			return is;
+		}
+		static void unlock(uint32_t is) {
+			Base::unlock();
+			__set_PRIMASK(is);
+		}
+	};
+
+	using SystemSpinLock = SpinLockBlocking<0>;
+
+	/**
+	 *
+	 * RAI lock acquiring and releasing
+	 */
+	template <class SpinLock>
+	class SpinLockGuard {
+	private:
+		SpinLockGuard(const SpinLockGuard& ) = delete;
+		SpinLockGuard(SpinLockGuard&&) = delete;
+	public:
+		SpinLockGuard() { SpinLock::lock(); }
+		~SpinLockGuard() { SpinLock::unlock(); }
+	};
+
+	template <uint8_t instance>
+	class SpinLockGuard<SpinLockBlocking<instance>> {
+	private:
+		using SpinLock = SpinLockBlocking<instance>;
+		uint32_t is;
+		SpinLockGuard(const SpinLockGuard& ) = delete;
+		SpinLockGuard(SpinLockGuard&&) = delete;
+	public:
+		SpinLockGuard() : is(SpinLock::lock()) {}
+		~SpinLockGuard() { SpinLock::unlock(is); }
+	};
+
+	using SystemSpinLockGuard = SpinLockGuard<SystemSpinLock>;
+
+	class Mutex {
+	private:
+		std::atomic<bool> locked = false;
```

Comment:
I just remembered that gcc doesn't implement C/C++ atomics for ARMv6-M (without load/store exclusive support) and leaves the specific implementation to the library.
I've implemented [atomics in modm via `modm::atomic::Lock`](https://github.com/modm-io/modm/blob/develop/ext/gcc/atomics_c11_cortex.cpp.in), however, this may not work correctly anymore for SMP?
Is there some addressing magic for atomic access on SRAM like for the peripherals?

Reply:
This variable (`locked`) accessed only on full hardware lock (spin locked another core and interrupts disabled), and there it is worked just as `volatile bool`

Reply:
I was thinking more generally about using `std::atomic` in user/library code to guard memory between cores/interrupts on this device, does that even work?

Reply:
I need to look more closely `modm::atomic ::Lock` looks like needed to add `SpinLock` to it.